### PR TITLE
Optimise vector allocations

### PR DESCRIPTION
**Pull request by SamuelMarks**
Updated with latest master and applied approach to more locations as #102868

If you like I can put it to one file per PR; as the TensorFlow and Keras people requested when I sent similar PRs.

### CODE REVIEW

File: c10/cuda/test/impl/CUDAAssertionsTest_multiple_writes_from_blocks_and_threads.cu [hunk 1]
```diff
@@ -52,7 +52,8 @@ void cuda_device_assertions_multiple_writes_from_blocks_and_threads() {
 
   // Spin up a bunch of busy-looping threads
   std::vector<std::thread> threads;
-  for (int i = 0; i < 10; i++) {
+  threads.reserve(10);
```
Comment (Skylion007) [inline c10/cuda/test/impl/CUDAAssertionsTest_multiple_writes_from_blocks_and_threads.cu:55]: NIT, this should probably moved into a variable as magic numbers are bad.

File: c10/test/util/irange_test.cpp [hunk 2]
```diff
@@ -17,6 +17,7 @@ TEST(irangeTest, range_test) {
 
 TEST(irangeTest, end_test) {
   std::vector<int> test_vec;
+  test_vec.reserve(5);
```
Comment (Skylion007) [inline c10/test/util/irange_test.cpp:20]: likewise here, the 5 should have 1 var that can be edited in the future.

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
Looks good but I have a few nits

**Comment by SamuelMarks** [comment]
@Skylion007 Arghhh hopefully I didn't scope creep with that latest commit. I addressed your nits also BTW

PS: I was looking at this and thinking to prereserve:
```c++
Module Module::clone(bool inplace) const {
  std::unordered_map<TypePtr, TypePtr> type_remap;
  IValue::HashAliasedIValueMap memo;
  const std::unordered_set<std::string> ignored_methods;
  const std::unordered_set<std::string> ignored_attributes;
  return clone_impl(
      type_remap, inplace, memo, ignored_methods, ignored_attributes);
}
```

**Comment by SamuelMarks** [comment]
Any reason this can't be merged @Skylion007 ?

**Comment by SamuelMarks** [comment]
@Skylion007 ?

**Pull request closed without merging**